config/internal/merge/merge_test.go

```diff
@@ -0,0 +1,516 @@
+				},
+			},
+			expected: &Example{
+				Struct: Struct{
+					IntField:    10,
+					StringField: "some string",
+				},
+			},
+		},
+		{
+			name: "dst(embedded_config)",
+			dst:  &Example{},
+			src: &map[string]any{
+				"embedded_int_field": 10,
+			},
+			expected: &Example{
+				EmbeddedStruct: EmbeddedStruct{
+					IntField: 10,
+				},
+			},
+		},
+		{
+			name: "src(embedded_config)",
+			dst: &Example{
+				EmbeddedStruct: EmbeddedStruct{
+					IntField: 10,
+				},
+			},
+			src: &map[string]any{},
+			expected: &Example{
+				EmbeddedStruct: EmbeddedStruct{
+					IntField: 10,
+				},
+			},
+		},
+		{
+			name: "overwrite(embedded_config)",
+			dst: &Example{
+				EmbeddedStruct: EmbeddedStruct{
+					IntField: 10,
+				},
+			},
+			src: &map[string]any{
+				"embedded_int_field": 20,
+			},
+			expected: &Example{
+				EmbeddedStruct: EmbeddedStruct{
+					IntField: 20,
+				},
+			},
+		},
+		{
+			name: "overwrite(1st_annotation)",
+			dst: &Example{
+				MultipleAnnotationField: 10,
+			},
+			src: &map[string]any{
+				"MultipleAnnotationField": 20,
+			},
+			expected: &Example{
+				MultipleAnnotationField: 20,
+			},
+		},
+		{
+			name: "overwrite(2nd_annotation)",
+			dst: &Example{
+				MultipleAnnotationField: 10,
+			},
+			src: &map[string]any{
+				"multiple_annocation_field": 20,
+			},
+			expected: &Example{
+				MultipleAnnotationField: 20,
+			},
+		},
+		{
+			name: "complex",
+			dst:  &Example{},
+			src: &map[string]any{
+				"struct": map[string]any{
+					"complex_map": map[string]any{
+						"complex_struct": map[string]any{
+							"complex_slice": []any{
+								map[string]any{
+									"simple_field": "value",
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: &Example{
+				Struct: Struct{
+					ComplexMap: map[string]ComplexStruct{
+						"complex_struct": {
+							ComplexSlice: []SimpleStruct{
+								{
+									SimpleField: "value",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:      "negative(src=nil)",
+			dst:       &Example{},
+			src:       nil,
+			expected:  &Example{},
+			expectErr: true,
+		},
+		{
+			name:      "negative(dst=nil)",
+			dst:       nil,
+			src:       &map[string]any{},
+			expectErr: true,
+		},
+		{
+			name: "negative(cast error)",
+			dst:  &Example{},
+			src: &map[string]any{
+				"int_field": "some string",
+			},
+			expected:  &Example{},
+			expectErr: true,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			err := Merge(test.dst, test.src)
+			if err != nil && !test.expectErr {
+				t.Fatalf("unexpected error: %v", err)
+			} else if err == nil && test.expectErr {
+				t.Fatal("expected error")
+			} else if !reflect.DeepEqual(test.expected, test.dst) {
+				t.Fatalf("invalid config.\nexpected: %v\n  actual: %v", test.expected, test.dst)
+			}
+		})
+	}
+
+	t.Run("negative(dst=value)", func(t *testing.T) {
+		if err := Merge(10, &map[string]any{}); err == nil {
+			t.Fatal("expected error")
```

Comment:
For negative test cases, we should also assert on the type of error we get back. It's easy to end up with bugs because you get an error, but not the one you expect.